### ISSUE TITLE
com.sun.jna.platform.win32.Win32Exception: The function requested is not supported

### ISSUE THREAD

**Issue by Veteraanikoodari** [Waffle/waffle#709]
Hello!

Description of environment:

We have a native JavaClient, WebClient and a our own JavaServer with built in WebServer (based on simpleframework.org) and everything works as expected with JavaClient. Both Negotiate and NTLM work as expected. 

Negotiate and NTLM also work with WebClient as long as JavaServer is running on Windows 10. However, if JavaServer is run on Windows Server 2012 R2 or Windows Server 2016 the Negotiate protocol from WebClient fails with: 

com.sun.jna.platform.win32.Win32Exception: The function requested is not supported

This happens when I'm calling WindowsAuthProviderImpl.acceptSecurityToken(String connectionId, byte[] token, String securityPackage) for the 2nd time. First call for that function succeeds but when the return ticket from the browser is supplied (Kerberos phase 2) I'm getting the aforementioned error.

1. If I downgrade the securityPackage to NTLM (for WebClient), everything works again on all platforms.
2. If JavaServer is run on Windows 10, everything works on all platforms.

The issue only exists for WebClient when JavaServer is running on Windows Server and the issue only exists for the second part of Negotiate protocol.

**Comment by hazendaz** [comment]
Hi,  which version of waffle are you using?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Veteraanikoodari** [comment]
We are using 1.9.0. Also tried with latest beta with same results.

**Comment by hazendaz** [comment]
Thanks for trying both.  A lot changed with how jna works with the latest beta.  I was planning to take it out of beta as working on various tests at this point successfully, however, I don't have a windows 2012 machine to work with.  Is it possible jna dropped such support?  You migth want to look at the jna  projecthttps://github.com/java-native-access/jna.  If support was indeed dropped, you can on 1.9.1 (bump up from 1.9.0 but no expected difference) drop back jna to any prior version in recent years and that should in theory work.  With the latest beta that option would not be possible as the underpinnings changed.  At the moment I'm not sure what the issue is but think these ideas might help you in possibly finding a solution there.  Although windows 2012 isn't all that old to start with...

**Comment by Veteraanikoodari** [comment]
The problem also exists with Windows Server 2016 as described. Based on our testing it is somehow related to how domains are identified between two different computers with different operating systems (Windows 10 and Server 2016).

Example: WebClient is running on Windows 10, JavaServer is running on Windows Server 2016:

Negotiate doesn't work, NTLM works. However, I can 'activate' Negotiate with IE, by entering the fully qualified domain name into the address bar. That causes the windows login dialog to appear, and when credentials are entered there, the Negotiate procedure executes successfully. So somehow the windows login dialog fetches/creates a different kind of Kerberos ticket than the browser does?

Maybe the server computer and client computer see domain differently (Fully qualified vs. shortcut) and Waffle or Kerberos doesn't accept that?

**Comment by pedroneil** [comment]
Hello,

As far as I am aware the kerberos configuration of your servers and SPN are
very specific. Are you using a server 2003 compatible domain or 2016
domain.  This also makes genesis and Negotiate differently.

With NTLM your server can query the domain
With negotiate / kerberos you must have an SPN configured for the domain
name and the domain must be exactly the same as domain of the server
hostname.

There are FAQ sections for setting up your spn

**Comment by Veteraanikoodari** [comment]
Thanks for taking the time to help!

1. We created the SPN(s) as instructed in FAQ
This led to a different kind of error, now the first token was accepted without continue?

2. I found out that our Negotiate process had (when working) defaulted to NTLM. All the examples I've come across described a two stage process which is not (usually) the case for Kerberos. So I needed to fix the server side of things to accept both one and two (mutual authentication) stage processes. 

Now we had both NTLM and Kerberos working. Only issue left was that Chrome and IE kept asking for credentials when fully qualified name for the server was used instead of short version.
 
3. The last problem was fixed by configuring fqdn in: 
Internet options -> Security -> Local intranet -> Sites -> Advanced -> Add

Now everything works as expected. Thank you very much for your input!
